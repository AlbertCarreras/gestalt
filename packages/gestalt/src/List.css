.bareList:first-child {
  margin: 0;
  padding-inline-start: 0;
}

.list {
  margin: 0;
  padding-inline-start: 16px;
}

.regular {
  margin-bottom: -16px;
  margin-top: -16px;
}

.condensed {
  margin: -8px;
  padding-inline-start: -8px;
}

.listItem {
  margin-bottom: 16px;
  margin-top: 16px;
}

.listItemCondensed {
  margin-bottom: 8px;
  margin-top: 8px;
}

.noStyle {
  list-style-type: none;
}

.ulItemDot {
  list-style-type: disc;
}

.ulItemCircle {
  list-style-type: circle;
}

.olItem1 {
  list-style-type: decimal;
}

.olItemA {
  list-style-type: upper-latin;
}

.olItema {
  list-style-type: lower-latin;
}

.textColorOverrideLight * {
  /* stylelint-disable-next-line declaration-no-important */
  color: var(--color-text-light) !important;
}

.textColorOverrideDark * {
  /* stylelint-disable-next-line declaration-no-important */
  color: var(--color-text-dark) !important;
}
