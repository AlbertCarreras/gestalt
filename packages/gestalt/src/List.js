// @flow strict
import { useId, type Node, type Element } from 'react';
import classnames from 'classnames';
import Box from './Box.js';
import Text from './Text.js';
import Flex from './Flex.js';
import ListItem from './ListItem.js';
import ListText from './ListText.js';
import NestedList from './NestedList.js';
import styles from './List.css';
import getChildrenToArray from './List/getChildrenToArray.js';
import { ListProvider } from './contexts/ListProvider.js';
import { NestingProvider } from './contexts/NestingProvider.js';

const UNORDERED_SEQUENCE = ['desc', 'circle'];
const ORDERED_SEQUENCE = ['decimal', 'upper-latin', 'lower-latin'];

// These sequences are used to style list-style-type in List.Item. Lists can mix ordered and unordered list. These sequences support carrying the item style independently from the nested level.
const STYLE_SEQUENCE_UNORDERED = Object.freeze([
  ...UNORDERED_SEQUENCE,
  ...UNORDERED_SEQUENCE,
  ...UNORDERED_SEQUENCE,
]);
const STYLE_SEQUENCE_ORDERED = Object.freeze([...ORDERED_SEQUENCE, ...ORDERED_SEQUENCE]);

type ListType = 'bare' | 'ordered' | 'unordered';

type Props = {|
  /**
   * Use List.Item to build lists. See [subcomponents](https://gestalt.pinterest.systems/web/list#Subcomponents).
   */
  children: Node,
  /**
   * The label for the list. Be sure to localize the text. See the [label variant](https://gestalt.pinterest.systems/web/list#Text-and-label) for guidance.
   */
  label: string | Element<typeof Text>,
  /**
   * Whether the label should be visible or not. If `hidden`, the label is still available for screen reader users, but does not appear visually. See the [accessibility section](https://gestalt.pinterest.systems/web/list#Accessibility) and the [label variant](https://gestalt.pinterest.systems/web/list#Text-and-label) for guidance.
   */
  labelDisplay?: 'visible' | 'hidden',
  /**
   * The spacing used in List. See the [spacing variant](https://gestalt.pinterest.systems/web/list#Spacing) for guidance.
   */
  spacing?: 'regular' | 'condensed',
  /**
   * Determines the style of the list. See the [type variant](https://gestalt.pinterest.systems/web/list#Type) to learn more.
   */
  type?: ListType,
|};

/**
 * [List](https://gestalt.pinterest.systems/web/list) allows users to view individual, but related, text items grouped together.
 * ![List light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/List.spec.mjs-snapshots/List-chromium-darwin.png)
 * ![List dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/List-dark.spec.mjs-snapshots/List-dark-chromium-darwin.png)
 */
function List({
  label,
  labelDisplay = 'visible',
  spacing = 'regular',
  type,
  children,
}: Props): Node {
  const id = useId();

  const listType = type ?? 'unordered';

  const hiddenLabel = labelDisplay === 'hidden';

  const ListElement = listType === 'ordered' ? 'ol' : 'ul';

  const listChildren = getChildrenToArray({ children, filterLevel: 'List' });

  const className = classnames(styles.list, {
    [styles.bareList]: type === 'bare',
    [styles.regular]: type === 'regular',
    [styles.condensed]: type === 'condensed',
  });

  return (
    <ListProvider
      type={listType}
      spacing={spacing}
      style={{ ol: STYLE_SEQUENCE_ORDERED, ul: STYLE_SEQUENCE_UNORDERED }}
    >
      <NestingProvider componentName="List" maxNestedLevels={6}>
        <Flex direction="column">
          <Box
            id={id}
            display={hiddenLabel ? 'visuallyHidden' : 'block'}
            marginBottom={hiddenLabel ? 0 : 3}
          >
            <ListText text={label} />
          </Box>
          <Text>
            <ListElement aria-labelledby={id} className={className}>
              {listChildren}
            </ListElement>
          </Text>
        </Flex>
      </NestingProvider>
    </ListProvider>
  );
}

List.displayName = 'List';

List.Item = ListItem;
List.NestedList = NestedList;

export default List;
