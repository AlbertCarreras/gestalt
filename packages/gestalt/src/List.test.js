// @flow strict
import { create } from 'react-test-renderer';
import List from './List.js';

describe('List', () => {
  it('renders', () => {
    const tree = create(<List name="" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an accessibility label', () => {
    const tree = create(<List accessibilityLabel="Test Accessibility Label" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
