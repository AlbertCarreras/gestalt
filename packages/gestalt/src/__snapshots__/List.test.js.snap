// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders a bare unordered list 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r0:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ul
      aria-labelledby=":r0:"
      className="list bareList"
    >
      <li
        className="noStyle listItem"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
        <ul
          className="list"
        >
          <li
            className="noStyle listItem"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
            <ul
              className="list"
            >
              <li
                className="noStyle listItem"
              >
                <div
                  className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                >
                  List item text
                </div>
                <ul
                  className="list"
                >
                  <li
                    className="noStyle listItem"
                  >
                    <div
                      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                    >
                      List item text
                    </div>
                    <ul
                      className="list"
                    >
                      <li
                        className="noStyle listItem"
                      >
                        <div
                          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                        >
                          List item text
                        </div>
                        <ul
                          className="list"
                        >
                          <li
                            className="noStyle listItem"
                          >
                            <div
                              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                            >
                              List item text
                            </div>
                          </li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;

exports[`List renders an condensed list 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r6:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r6:"
      className="list"
    >
      <li
        className="listItemCondensed olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an list item with custom text 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r7:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r7:"
      className="list"
    >
      <li
        className="listItemCondensed olItem1"
      >
        <span
          className="textColorOverrideDark"
        >
          <div
            className="Text fontSize300 defaultText alignStart breakWord fontWeightSemiBold"
          >
            List item text
          </div>
        </span>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an list with a custom label 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r4:"
  >
    <span
      className="textColorOverrideDark"
    >
      <div
        className="Text fontSize300 defaultText alignStart breakWord fontWeightSemiBold"
      >
        label test
      </div>
    </span>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r4:"
      className="list"
    >
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an list with a hidden label 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayVisuallyHidden"
    id=":r5:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r5:"
      className="list"
    >
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an mixed list 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r3:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r3:"
      className="list"
    >
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
      </li>
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
        <ul
          className="list"
        >
          <li
            className="listItem ulItemDot"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
          </li>
          <li
            className="listItem ulItemDot"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
            <ul
              className="list"
            >
              <li
                className="listItem ulItemCircle"
              >
                <div
                  className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                >
                  List item text
                </div>
              </li>
              <li
                className="listItem ulItemCircle"
              >
                <div
                  className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                >
                  List item text
                </div>
                <ol
                  className="list"
                >
                  <li
                    className="listItem olItem1"
                  >
                    <div
                      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                    >
                      List item text
                    </div>
                  </li>
                  <li
                    className="listItem olItem1"
                  >
                    <div
                      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                    >
                      List item text
                    </div>
                  </li>
                </ol>
              </li>
            </ul>
          </li>
          <li
            className="listItem ulItemDot"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
          </li>
        </ul>
      </li>
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an ordered list 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r2:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ol
      aria-labelledby=":r2:"
      className="list"
    >
      <li
        className="listItem olItem1"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
        <ol
          className="list"
        >
          <li
            className="listItem olItemA"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
            <ol
              className="list"
            >
              <li
                className="listItem olItema"
              >
                <div
                  className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                >
                  List item text
                </div>
                <ol
                  className="list"
                >
                  <li
                    className="listItem olItem1"
                  >
                    <div
                      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                    >
                      List item text
                    </div>
                    <ol
                      className="list"
                    >
                      <li
                        className="listItem olItemA"
                      >
                        <div
                          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                        >
                          List item text
                        </div>
                        <ol
                          className="list"
                        >
                          <li
                            className="listItem olItema"
                          >
                            <div
                              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                            >
                              List item text
                            </div>
                          </li>
                        </ol>
                      </li>
                    </ol>
                  </li>
                </ol>
              </li>
            </ol>
          </li>
        </ol>
      </li>
    </ol>
  </div>
</div>
`;

exports[`List renders an unordered list 1`] = `
<div
  className="Flex rowGap0 columnGap0 xsDirectionColumn"
>
  <div
    className="box xsDisplayBlock"
    id=":r1:"
  >
    <div
      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
    >
      label test
    </div>
  </div>
  <div
    className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
  >
    <ul
      aria-labelledby=":r1:"
      className="list"
    >
      <li
        className="listItem ulItemDot"
      >
        <div
          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
        >
          List item text
        </div>
        <ul
          className="list"
        >
          <li
            className="listItem ulItemCircle"
          >
            <div
              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
            >
              List item text
            </div>
            <ul
              className="list"
            >
              <li
                className="listItem ulItemDot"
              >
                <div
                  className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                >
                  List item text
                </div>
                <ul
                  className="list"
                >
                  <li
                    className="listItem ulItemCircle"
                  >
                    <div
                      className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                    >
                      List item text
                    </div>
                    <ul
                      className="list"
                    >
                      <li
                        className="listItem ulItemDot"
                      >
                        <div
                          className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                        >
                          List item text
                        </div>
                        <ul
                          className="list"
                        >
                          <li
                            className="listItem ulItemCircle"
                          >
                            <div
                              className="Text fontSize300 defaultText alignStart breakWord fontWeightNormal"
                            >
                              List item text
                            </div>
                          </li>
                        </ul>
                      </li>
                    </ul>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</div>
`;
